compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala

```diff
@@ -0,0 +1,451 @@
+      |newSource1.scala:7: error: js.native may only be used as stub implementation in facade types
+      |        def a: Int = js.native
+      |                        ^
+    """
+  }
+
+  @Test
+  def noNonLiteralJSName: Unit = {
+    """
+    object A {
+      val a = "Hello"
+      final val b = "World"
+    }
+
+    @ScalaJSDefined
+    class B extends js.Object {
+      @JSName(A.a)
+      def foo: Int = js.native
+      @JSName(A.b)
+      def bar: Int = js.native
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:12: error: The argument to JSName must be a literal string
+      |      @JSName(A.a)
+      |       ^
+    """
+  }
+
+  @Test
+  def noApplyProperty: Unit = {
+    // def apply
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      def apply: Int = 42
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:7: error: A member named apply represents function application in JavaScript. A parameterless member should be exported as a property. You must add @JSName("apply")
+      |      def apply: Int = 42
+      |          ^
+    """
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      @JSName("apply")
+      def apply: Int = 42
+    }
+    """.succeeds
+
+    // val apply
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      val apply: Int = 42
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:7: error: A member named apply represents function application in JavaScript. A parameterless member should be exported as a property. You must add @JSName("apply")
+      |      val apply: Int = 42
+      |          ^
+    """
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      @JSName("apply")
+      val apply: Int = 42
+    }
+    """.succeeds
+
+    // var apply
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      var apply: Int = 42
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:7: error: A member named apply represents function application in JavaScript. A parameterless member should be exported as a property. You must add @JSName("apply")
+      |      var apply: Int = 42
+      |          ^
+    """
+
+    """
+    @ScalaJSDefined
+    class A extends js.Object {
+      @JSName("apply")
+      var apply: Int = 42
+    }
+    """.succeeds
+  }
+
+  @Test
+  def noExportClass: Unit = {
+    """
+    @ScalaJSDefined
+    @JSExport
+    class A extends js.Object
+    """ hasErrors
+    """
+      |newSource1.scala:6: error: Implementation restriction: exporting a Scala.js-defined JS class is not supported
+      |    @JSExport
+      |     ^
+    """
+  }
+
+  @Test
+  def noConcreteMemberInTrait: Unit = {
+    """
+    @ScalaJSDefined
+    trait A extends js.Object {
+      def foo(x: Int): Int = x + 1
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:7: error: A Scala.js-defined JS trait can only contain abstract members
+      |      def foo(x: Int): Int = x + 1
+      |          ^
+    """
+  }
```

Comment:
Ah, yes, I saw it. The error message is a bit convoluted, but that's fine. No need to complicate compiler code. Who puts private methods in traits you cannot put implementations in anyways...